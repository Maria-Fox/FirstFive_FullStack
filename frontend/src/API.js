import axios from "axios";

const BASE_URL = process.env.REACT_APP_BASE_URL || "https://localhost:3001";

class API {

  // Used to authenticate user
  static token = "";

  // ******************************************* API send request method

  static async request(endpoint, data = {}, method = "get") {
    // console.debug("API Call:", endpoint, data, method, this.token, "**********");

    if (!this.token) {
      this.token = window.localStorage.getItem("token") || "invalidTemp"
    };

    const url = `${BASE_URL}/${endpoint}`;
    const headers = { Authorization: `Bearer ${API.token}`, 'Access-Control-Allow-Origin': 'https://firstfive.onrender.comtps://firstfive.surge.sh'};
    const params = (method === "get")
      ? data
      : {};


    try {
      return (await axios({ url, method, data, params, headers })).data;
    } catch (err) {
      // console.error("API Error:", err.message);
      // grab just the error message
      let message = err.response.data.error.message;
      throw Array.isArray(message) ? message : [message];
    };
  };

  // ******************************************* Register/ LOGIN methods


  static async registerUser(userData) {
    let newUser = await this.request("auth/register", { ...userData }, 'post');
    this.token = newUser.signedJWT
    return newUser.signedJWT;
  };

  static async authenticateUser(userData) {
    let authUser = await this.request(`auth/login`, { ...userData }, 'post');
    return authUser.signedJWT;
  };




  // ******************************************* USER methods

  // Need to see about limmited views to only those who matched..


  static async viewUser(username) {
    let res = await this.request(`users/${username}`);
    return res.userData;
  };

  static async viewAuthUserProfile(username) {
    let res = await this.request(`users/${username}/profile`);
    return res.userData;
  };

  static async editUser(username, updateData) {
    let res = await this.request(`users/${username}`, { ...updateData }, 'patch');
    return res.userData;
  };

  static async deleteUser(username) {
    let res = await this.request(`users/${username}`, {}, 'delete');
    return res;
  };


  // ******************************************* PROJECT methods

  static async createProject(projectData) {
    let res = await this.request(`projects/add`, { ...projectData }, 'post');
    return res;
  };

  static async getAllProjects() {
    let res = await this.request(`projects/all`);
    return res;
  };

  static async getNonMatchedProjects() {
    let res = await this.request(`projects/unmatched`);
    return res;
  };

  static async carouselProjects() {
    let res = await this.request(`projects/carousel`);
    return res;
  };

  static async getUserCreatedProjects(username) {
    let res = await this.request(`projects/created/by/${username}`);
    return res;
  };

  static async viewProject(project_id) {
    let res = await this.request(`projects/${project_id}`);
    return res;
  };

  static async editProject(project_id, updateData) {
    let res = await this.request(`projects/${project_id}`, { ...updateData }, 'patch');
    return res;
  };

  static async deleteProject(project_id) {
    let res = await this.request(`projects/${project_id}`, {}, 'delete');
    return res;
  };

  // ******************************************* Match methods

  static async addMatch(username, project_id) {
    let res = await this.request(`matches/${username}/${project_id}`, {}, 'post');
    return res;
  };

  static async viewUserMatches(username) {
    let res = await this.request(`matches/${username}/all`);
    return res;
  };

  static async viewProjectUserMatches(project_id) {
    let res = await this.request(`matches/${project_id}/users`);
    return res;
  };

  static async removeUserMatch(username, project_id) {
    let res = await this.request(`matches/${username}/${project_id}`, {}, 'delete');
    return res;
  };


  // ******************************************* Project Member methods

  static async addProjectMember(project_id, username) {
    let res = await this.request(`projectmembers/add/${project_id}`, { username }, 'post');
    return res;
  };

  static async viewAllProjMembers(project_id) {
    let res = await this.request(`projectmembers/${project_id}/users`);
    return res;
  };

  static async deleteProjectMember(project_id, username) {
    let res = await this.request(`projectmembers/${project_id}`, { username }, 'delete');
    return res;
  };

  // ******************************************* Message methods

  static async createMessage(username, msgData) {
    let res = await this.request(`messages/${username}/create`, { ...msgData }, 'post');
    return res;
  };

  static async getAllUserMessages(username) {
    let res = await this.request(`messages/${username}/all`);
    return res;
  };

  static async viewSingleMsgData(message_id, username) {
    let res = await this.request(`messages/${username}/read/${message_id}`);
    return res;
  };


  // class closing bracket don't delete.
};

export default API;