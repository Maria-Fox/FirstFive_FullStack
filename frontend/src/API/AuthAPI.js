import axios from "axios";

const BASE_URL = process.env.REACT_APP_BASE_URL || "http://localhost:3001";

class AuthAPI {

  // Used to authenticate user
  static token = "";

  // ******************************************* API send request method

  static async request(endpoint, data = {}, method = "get") {

    if (!this.token) {
      this.token = window.localStorage.getItem("token") || "invalidTemp"
    };

    const url = `${BASE_URL}/${endpoint}`;
    const headers = { Authorization: `Bearer ${API.token}` };
    const params = (method === "get")
      ? data
      : {};


    try {
      return (await axios({ url, method, data, params, headers })).data;
    } catch (err) {
      console.error("API Error:", err.message);
      // grab just the error message
      let message = err.response.data.error.message;
      throw Array.isArray(message) ? message : [message];
    };
  };


  // ******************************************* Register/ LOGIN methods


  static async registerUser(userData) {
    let newUser = await this.request("auth/register", { ...userData }, 'post');
    this.token = newUser.signedJWT
    return newUser.signedJWT;
  };

  static async authenticateUser(userData) {
    let authUser = await this.request(`auth/login`, { ...userData }, 'post');
    return authUser.signedJWT;
  };

};

export default AuthAPI;