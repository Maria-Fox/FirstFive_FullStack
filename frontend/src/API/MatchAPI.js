import axios from "axios";

const BASE_URL = process.env.REACT_APP_BASE_URL || "http://localhost:3001";

class MatchAPI {

  // Used to authenticate user. Sent in request header.
  static token = "";

  // ******************************************* API send request method

  static async request(endpoint, data = {}, method = "get") {

    if (!this.token) {
      this.token = window.localStorage.getItem("token") || "invalidTemp"
    };

    const url = `${BASE_URL}/${endpoint}`;
    const headers = { Authorization: `Bearer ${API.token}` };
    const params = (method === "get")
      ? data
      : {};


    try {
      return (await axios({ url, method, data, params, headers })).data;
    } catch (err) {
      console.error("API Error:", err.message);
      // grab just the error message
      let message = err.response.data.error.message;
      throw Array.isArray(message) ? message : [message];
    };
  };


  // ******************************************* Match methods

  static async addMatch(username, project_id) {
    let res = await this.request(`matches/${username}/${project_id}`, {}, 'post');
    return res;
  };

  static async viewUserMatches(username) {
    let res = await this.request(`matches/${username}/all`);
    return res;
  };

  static async removeUserMatch(username, project_id) {
    let res = await this.request(`matches/${username}/${project_id}`, {}, 'delete');
    return res;
  };

  static async viewProjectUserMatches(project_id) {
    let res = await this.request(`matches/${project_id}/users`);
    return res;
  };


};

export default MatchAPI;