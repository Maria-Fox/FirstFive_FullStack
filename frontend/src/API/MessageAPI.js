import axios from "axios";

const BASE_URL = process.env.REACT_APP_BASE_URL || "http://localhost:3001";

class MessageAPI {

  // Used to authenticate user
  static token = "";

  // ******************************************* API send request method

  static async request(endpoint, data = {}, method = "get") {

    if (!this.token) {
      this.token = window.localStorage.getItem("token") || "invalidTemp"
    };

    const url = `${BASE_URL}/${endpoint}`;
    const headers = { Authorization: `Bearer ${API.token}` };
    const params = (method === "get")
      ? data
      : {};


    try {
      return (await axios({ url, method, data, params, headers })).data;
    } catch (err) {
      console.error("API Error:", err.message);
      // grab just the error message
      let message = err.response.data.error.message;
      throw Array.isArray(message) ? message : [message];
    };
  };


  // ******************************************* Message methods

  static async createMessage(username, msgData) {
    let res = await this.request(`messages/${username}/create`, { ...msgData }, 'post');
    return res;
  };

  static async getAllUserMessages(username) {
    let res = await this.request(`messages/${username}/all`);
    return res;
  };

  static async viewSingleMsgData(message_id, username) {
    let res = await this.request(`messages/${username}/read/${message_id}`);
    return res;
  };

};

export default MessageAPI;