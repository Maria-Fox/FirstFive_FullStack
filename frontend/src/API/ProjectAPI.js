import axios from "axios";

const BASE_URL = process.env.REACT_APP_BASE_URL || "http://localhost:3001";

class ProjectAPI {

  // Used to authenticate user
  static token = "";

  // ******************************************* API send request method

  static async request(endpoint, data = {}, method = "get") {

    if (!this.token) {
      this.token = window.localStorage.getItem("token") || "invalidTemp"
    };

    const url = `${BASE_URL}/${endpoint}`;
    const headers = { Authorization: `Bearer ${API.token}` };
    const params = (method === "get")
      ? data
      : {};


    try {
      return (await axios({ url, method, data, params, headers })).data;
    } catch (err) {
      console.error("API Error:", err.message);
      // grab just the error message
      let message = err.response.data.error.message;
      throw Array.isArray(message) ? message : [message];
    };
  };


  // ******************************************* PROJECT methods

  static async createProject(projectData) {
    let res = await this.request(`projects/add`, { ...projectData }, 'post');
    return res;
  };

  static async getAllProjects() {
    let res = await this.request(`projects/all`);
    return res;
  };

  static async getNonMatchedProjects() {
    let res = await this.request(`projects/unmatched`);
    return res;
  };

  static async carouselProjects() {
    let res = await this.request(`projects/carousel`);
    return res;
  };

  static async getUserCreatedProjects(username) {
    let res = await this.request(`projects/created/by/${username}`);
    return res;
  };

  static async viewProject(project_id) {
    let res = await this.request(`projects/${project_id}`);
    return res;
  };

  static async editProject(project_id, updateData) {
    let res = await this.request(`projects/${project_id}`, { ...updateData }, 'patch');
    return res;
  };

  static async deleteProject(project_id) {
    let res = await this.request(`projects/${project_id}`, {}, 'delete');
    return res;
  };

};

export default ProjectAPI;