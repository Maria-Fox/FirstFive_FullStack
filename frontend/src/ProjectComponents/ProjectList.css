#Project-links{
  color: whitesmoke;
  text-decoration: none;
}

/* #Project-links:hover{
  background-color: aquamarine;
  color: #000000e7;
  border-radius: 10px;
}; */