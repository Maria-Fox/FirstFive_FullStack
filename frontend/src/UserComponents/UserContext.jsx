import React from "react";

// Create user context instance to later provide in App.js and use throughout application. Holds auth user

const UserContext = React.createContext();

export default UserContext;