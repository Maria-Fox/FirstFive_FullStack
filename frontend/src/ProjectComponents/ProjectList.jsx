import React, { useContext, useEffect, useState } from "react";
import { Link } from "react-router-dom";
import API from "../API";
import UserContext from "../UserComponents/UserContext";
import ProjectCard from "./ProjectCard";
import confetti from "canvas-confetti";
import AlertNotification from "../Common/AlertNotifications";
import { Card, CardText } from "reactstrap";
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {faEdit} from "@fortawesome/free-regular-svg-icons";
import { regular } from '@fortawesome/fontawesome-svg-core/import.macro'; // <-- import styles to be used
import "./ProjectList.css";


const ProjectList = () => {

  // ***************************************************************


  let {authUser, matchedProjectIds, setMatchedProjectIds } = useContext(UserContext);

  let [projects, setProjects] = useState(null);
  const [errors, setErrors] = useState(null);

  // ***************************************************************

  useEffect(function viewAllProjects() {
    async function getAllProjects() {

      try {
        if (matchedProjectIds.length === 0) {
          let response = await API.getAllProjects();
          setProjects(response);
        } else {
          let response = await API.getNonMatchedProjects();
          setProjects(response);
        }
      } catch (e) {
        setErrors(e);
        return;
      };
    };

    getAllProjects();

  }, [setMatchedProjectIds, setProjects, authUser]);

  // ***************************************************************


  let handleMatch = async function (authUser, id) {
    try {
      await API.addMatch(authUser, id);

      // Add the project to matches.
      setMatchedProjectIds(matchedProjectIds => [...matchedProjectIds, id]);

      // Update displayed unmatched projects, removing the latest match.
      setProjects(projects.filter(projects => projects.id !== id));
      confetti();
    } catch (e) {
      setErrors(e);
      return;
    };
  };

  // ***************************************************************

  let noProjectsToMatch = (
    <Card className="m-3 p-3">
      <CardText style={{ textAlign: "center" }}>No projects left to match. Come back and check again soon!</CardText>
    </Card>
  );

  // ***************************************************************

  return (
    <div className="container">
      <h1 className="text-center text-white pt-2 mt-2">Projects</h1>

      {errors ? <AlertNotification messages={errors} /> : null}

      <Card className="container pt-3 pb-3 mt-2 mb-4 bg-dark bg-gradient text-white d-flex justify-content-between">
        <CardText >

          <Link to="/projects/carousel" 
          id= "Project-links" className = "d-flex justify-content-between p-2"
          style = {{color: "whitesmoke", textDecoration: "none"}}
          >
            View the projects through match-cards <FontAwesomeIcon icon={regular('heart')} />
          </Link>

          <Link to="/projects/add" id= "Project-links" 
          className = "d-flex justify-content-between p-2"
          style = {{color: "whitesmoke", textDecoration: "none"}}>
            Create a project <FontAwesomeIcon icon={faEdit} /> 
          </Link>

        </CardText>
      </Card>

      {
        projects && projects.length > 0 ? projects.map(({ id, owner_username, name, project_desc, timeframe, github_repo }) =>
          <ProjectCard
            key={id}
            id={id}
            owner_username={owner_username}
            name={name}
            project_desc={project_desc}
            timeframe={timeframe}
            github_repo={github_repo}
            handleMatch={handleMatch}
          />
        )
          :
          noProjectsToMatch
      }
    </div >
  );
};

export default ProjectList;