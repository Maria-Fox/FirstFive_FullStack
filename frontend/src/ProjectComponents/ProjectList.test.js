import { render, getByText, waitFor } from "@testing-library/react";
import ProjectList from "./ProjectList";
import { MemoryRouter } from "react-router-dom";
import UserContext from "../UserComponents/UserContext";
// Mock API calls.
import API from "../API";
jest.mock("../API")


describe("ProjectList component- approp display", function () {

  test("Smoke - Render ProjectList component", function () {

    const validUser = { username: "softwareDev1", matchedProjectIds: [] };

    render(
      <UserContext.Provider value={validUser}>
        <MemoryRouter>
          <ProjectList />
        </MemoryRouter>
      </UserContext.Provider>
    )
  });


  test("Valid user, renders no project to display if no API call yet.", function () {

    const validUser = { username: "softwareDev1", matchedProjectIds: [] };

    //Not mocking API call- instead, disaplying no Projects

    const { asFragment, getByText } = render(
      <MemoryRouter>
        <UserContext.Provider value={validUser}>
          <ProjectList />
        </UserContext.Provider>
      </MemoryRouter >
    );

    expect(getByText('Projects')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });


  test("Valid user, renders full project list WITH matchedprojects.", function () {

    const validUserContext = { username: "softwareDev1", matchedProjectIds: [] };

    API.mockImplementationOnce(() => {
      return {
        request: () => {
          [{
            id: 4,
            owner_username: "Delia_Designs",
            name: "Design a smart TV APP layout",
            project_desc: "Beautifully design an interface for smart TV platforms. Include comprehnsive testing for all devices.",
            timeframe: "3-6 months",
            github_repo: "https://github.com/"
          }]
        }
      }
    });

    const renderedBody = render(
      <UserContext.Provider value={validUserContext} >
        <MemoryRouter>
          <ProjectList />
        </MemoryRouter>
      </UserContext.Provider >
    )

    waitFor(() => {
      expect(renderedBody.getByText("Design a smart TV APP layout")).toBeInTheDocument();
      expect(renderedBody.getByText("3-6months")).toBeInTheDocument();
    });
  });


});