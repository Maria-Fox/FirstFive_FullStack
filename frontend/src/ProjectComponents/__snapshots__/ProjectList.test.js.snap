// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectList component- approp display Valid user, renders no project to display if no API call yet. 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <h1
      class="text-center text-white pt-2 mt-2"
    >
      Projects
    </h1>
    <div
      class="container pt-3 mt-2 mb-4 bg-dark bg-gradient text-white d-flex justify-content-between card"
    >
      <p
        class="card-text"
      >
        <a
          class="d-flex justify-content-between p-2"
          href="/projects/carousel"
          id="Project-links"
        >
          View the projects through match-cards 
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-heart "
            data-icon="heart"
            data-prefix="far"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M244 84L255.1 96L267.1 84.02C300.6 51.37 347 36.51 392.6 44.1C461.5 55.58 512 115.2 512 185.1V190.9C512 232.4 494.8 272.1 464.4 300.4L283.7 469.1C276.2 476.1 266.3 480 256 480C245.7 480 235.8 476.1 228.3 469.1L47.59 300.4C17.23 272.1 0 232.4 0 190.9V185.1C0 115.2 50.52 55.58 119.4 44.1C164.1 36.51 211.4 51.37 244 84C243.1 84 244 84.01 244 84L244 84zM255.1 163.9L210.1 117.1C188.4 96.28 157.6 86.4 127.3 91.44C81.55 99.07 48 138.7 48 185.1V190.9C48 219.1 59.71 246.1 80.34 265.3L256 429.3L431.7 265.3C452.3 246.1 464 219.1 464 190.9V185.1C464 138.7 430.4 99.07 384.7 91.44C354.4 86.4 323.6 96.28 301.9 117.1L255.1 163.9z"
              fill="currentColor"
            />
          </svg>
        </a>
        <a
          class="d-flex justify-content-between p-2"
          href="/projects/add"
          id="Project-links"
        >
          Create a project 
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-pen-to-square "
            data-icon="pen-to-square"
            data-prefix="far"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M441 58.9L453.1 71c9.4 9.4 9.4 24.6 0 33.9L424 134.1 377.9 88 407 58.9c9.4-9.4 24.6-9.4 33.9 0zM209.8 256.2L344 121.9 390.1 168 255.8 302.2c-2.9 2.9-6.5 5-10.4 6.1l-58.5 16.7 16.7-58.5c1.1-3.9 3.2-7.5 6.1-10.4zM373.1 25L175.8 222.2c-8.7 8.7-15 19.4-18.3 31.1l-28.6 100c-2.4 8.4-.1 17.4 6.1 23.6s15.2 8.5 23.6 6.1l100-28.6c11.8-3.4 22.5-9.7 31.1-18.3L487 138.9c28.1-28.1 28.1-73.7 0-101.8L474.9 25C446.8-3.1 401.2-3.1 373.1 25zM88 64C39.4 64 0 103.4 0 152V424c0 48.6 39.4 88 88 88H360c48.6 0 88-39.4 88-88V312c0-13.3-10.7-24-24-24s-24 10.7-24 24V424c0 22.1-17.9 40-40 40H88c-22.1 0-40-17.9-40-40V152c0-22.1 17.9-40 40-40H200c13.3 0 24-10.7 24-24s-10.7-24-24-24H88z"
              fill="currentColor"
            />
          </svg>
           
        </a>
      </p>
    </div>
    <div
      class="m-3 p-3 card"
    >
      <p
        class="card-text"
        style="text-align: center;"
      >
        No projects left to match. Come back and check again soon!
      </p>
    </div>
  </div>
</DocumentFragment>
`;
